include/pybind11/detail/common.h

```diff
@@ -154,6 +154,8 @@
 // C4505: 'PySlice_GetIndicesEx': unreferenced local function has been removed (PyPy only)
 #  pragma warning(disable: 4505)
 #  if defined(_DEBUG) && !defined(Py_DEBUG)
+// fix for vs2022: https://github.com/pybind/pybind11/issues/3477
+#    include <crtdefs.h>
```

Comment:
This seems to be causing issues for older MSVCs so I think the ifdef need to be more specific for new MSVCs if you want this work.

Reply:
I believe this introduces a new dependency on an optional component of MSVC / Windows.

Reply:
Since it's a specific case, it's probably okay (and might not be true, anyway). How about gating it for MSVC 2022+? We could probably also add `_CRT_SECURE_NO_WARNINGS` on Python 2.7, I hope people are not mixing 2.7 and MSVC 2022. Actually I hope people are not using 2.7 on Windows at all. Or at all.

Reply:
Yes, defining _CRT_SECURE_NO_WARNINGS will fix this problem.

I think it's a good idea to always define _CRT_SECURE_NO_WARNINGS, for all MSVC versions and all python versions. These warnings (treated by this build as errors) complain about use of standard C library functions and direct users toward non-standard Microsoft functions, which are not present on other platforms.

Reply:
> I think it's a good idea

Yes, but we shouldn't mess with other people's code. Using pybind11 should be as light weight as possible. If the compiler wants to tell the user something, we shouldn't get in the way, unless it's about pybind11 itself. The fewer surprises, the better. Every time someone tries to be "helpful" with something like this in a library, that causes issues down the road. Applications can do this, we could even suggest it, but we should not "just do it" for everyone.

Reply:
Then maybe we could stop treating these warnings as errors?

Reply:
I'm perfectly fine to hide these warnings inside pybind11, highly recommended actually, that's why we make the warnings errors, so we fix them. And we don't treat warnings as errors for anyone else.

(I honestly don't care enough about Python 2 to worry about leaking this if it's hard do to, by the way)

Reply:
Would it work to `#define _CRT_SECURE_NO_WARNINGS` only in this context, and only for VS2022? (Maybe that was suggested already?)
I think that would be an OK tradeoff.

In any case, I'd find it useful to generate a message like "INFO: Mixing non-debug Python build with debug pybind11 build. Please see *section in pybind11 docs* for background."
Ideally with added "_CRT_SECURE_NO_WARNINGS disabled" in this particular situation.
Although I have no idea how easy this is to implement.

Re "OK tradeoff":
* Clearly, playing tricks with a symbol that has a leading underscore would make any purist frown. We're obviously in "hack" territory right there.
* I'm strongly in favor of it through, as a matter of practicality.
* In that hack territory, it is nice if we can keep up as much code health tooling as possible, but if we run into a conflict between mixing debug modes and enabling warnings, I come down on the side of disabling the warnings that trip us up. Ultimately code that matters in real life will be built in release mode, the warnings will show up then.

Reply:
The warning is just due to some weird old CPython 2 stuff. The problem here is unrelated; we are intentionally going against the CPython requirement to try to hack around something that changed in MSVC 2022, this hack requires Python.h to be included before any system headers, directly in opposition to the docs.

Ideally:
* CPython should workaround this (if we can make a MWE, which is not too hard, I'd think, it would be a good report to make)
* MSVC could possibly do something to ensure this worked like before?
* This workaround (scoped as closely as possible) is fine for now, but could be broken by a future MSVC or CPython release, as neither supports this at all right now.

Reply:
I guess I'm still a bit lost, and I forgot to mention before, I'd `#ifdef` guard any workarounds also with `_MSC_VER`, to not change or break anything existing.

Henry wrote:
> this hack requires Python.h to be included before any system headers, directly in opposition to the docs.

Did you mean "this hack requires a system header to be included before Python.h, directly in opposition to the docs."

Meta... I just realized we have not two but three open threads:

#3477 an issue by @gearslogy
#3480 a PR by @henryiii 
#3497 a PR by @Boris-Rasin 

all about VS 2022. No wonder I'm confused :-) Please, let's close all but #3480 (maybe change the title) and work together / discuss there.